Modules/blake2module.c

```diff
@@ -0,0 +1,924 @@
+        PyMutex_Unlock(&self->mutex);
+        Py_END_ALLOW_THREADS
+    } else {
+        update(self, buf.buf, buf.len);
+    }
+
+    PyBuffer_Release(&buf);
+
+    Py_RETURN_NONE;
+}
+
+/*[clinic input]
+_blake2.blake2b.digest
+
+Return the digest value as a bytes object.
+[clinic start generated code]*/
+
+static PyObject *
+_blake2_blake2b_digest_impl(Blake2Object *self)
+/*[clinic end generated code: output=31ab8ad477f4a2f7 input=7d21659e9c5fff02]*/
+{
+    uint8_t digest[HACL_HASH_BLAKE2B_OUT_BYTES];
+
+    ENTER_HASHLIB(self);
+    uint8_t digest_length = 0;
+    switch (self->impl) {
+#if HACL_CAN_COMPILE_SIMD256
+        case Blake2b_256:
+            digest_length = Hacl_Hash_Blake2b_Simd256_digest(self->blake2b_256_state, digest);
+            break;
+#endif
+#if HACL_CAN_COMPILE_SIMD128
+        case Blake2s_128:
+            digest_length = Hacl_Hash_Blake2s_Simd128_digest(self->blake2s_128_state, digest);
+            break;
+#endif
+        case Blake2b:
+            digest_length = Hacl_Hash_Blake2b_digest(self->blake2b_state, digest);
+            break;
+        case Blake2s:
+            digest_length = Hacl_Hash_Blake2s_digest(self->blake2s_state, digest);
+            break;
+        default:
+            Py_UNREACHABLE();
+    }
+    LEAVE_HASHLIB(self);
+    return PyBytes_FromStringAndSize((const char *)digest, digest_length);
+}
+
+/*[clinic input]
+_blake2.blake2b.hexdigest
+
+Return the digest value as a string of hexadecimal digits.
+[clinic start generated code]*/
+
+static PyObject *
+_blake2_blake2b_hexdigest_impl(Blake2Object *self)
+/*[clinic end generated code: output=5ef54b138db6610a input=76930f6946351f56]*/
+{
+    uint8_t digest[HACL_HASH_BLAKE2B_OUT_BYTES];
+
+    ENTER_HASHLIB(self);
+    uint8_t digest_length = 0;
+    switch (self->impl) {
+#if HACL_CAN_COMPILE_SIMD256
+        case Blake2b_256:
+            digest_length = Hacl_Hash_Blake2b_Simd256_digest(self->blake2b_256_state, digest);
+            break;
+#endif
+#if HACL_CAN_COMPILE_SIMD128
+        case Blake2s_128:
+            digest_length = Hacl_Hash_Blake2s_Simd128_digest(self->blake2s_128_state, digest);
+            break;
+#endif
+        case Blake2b:
+            digest_length = Hacl_Hash_Blake2b_digest(self->blake2b_state, digest);
+            break;
+        case Blake2s:
+            digest_length = Hacl_Hash_Blake2s_digest(self->blake2s_state, digest);
+            break;
+        default:
+            Py_UNREACHABLE();
+    }
+    LEAVE_HASHLIB(self);
+    return _Py_strhex((const char *)digest, digest_length);
+}
+
+
+static PyMethodDef py_blake2b_methods[] = {
+    _BLAKE2_BLAKE2B_COPY_METHODDEF
+    _BLAKE2_BLAKE2B_DIGEST_METHODDEF
+    _BLAKE2_BLAKE2B_HEXDIGEST_METHODDEF
+    _BLAKE2_BLAKE2B_UPDATE_METHODDEF
+    {NULL, NULL}
+};
+
+
+static PyObject *
+py_blake2b_get_name(Blake2Object *self, void *closure)
+{
+    return PyUnicode_FromString(is_blake2b(self->impl) ? "blake2b" : "blake2s");
```

Comment:
Not sure if it's worth interning "blake2b" and "blake2s" strings.

Reply:
we don't bother in sha2module.c so lets not here either.